WIP on #293 - extended properties for discussion

FraserGreenroyd opened the pull request:
### Issues addressed by this PR
Fixes #293 

Begins work to include extended properties into base BHoM object for all BHoM objects to then make use of as we look to use extended properties on some objects. This PR has been raised to begin the review/discussion of how to implement this and showcase the ideas following discussions with @al-fisher  and @rwemay .


### Test files
N/A

### Changelog
####Added
 - Added `ExtendedProperties` functionality

adecler commented:
Glad to see some action happening around ECS 👍 .
A few initial comments:
- I am not convinced we need an `ElementID` property (especially a string) and therefore a `BHoMExtendedProperties` class (an interface would suffice).
- Names like `BHoMExtendedProperties` and `BuildingElementExtendedProperties` are really long and technical. I wonder if we could use something a bit less so. If you look at the `Human_oM` ([here](https://github.com/BHoM/BHoM/blob/master/Humans_oM/Interfaces/IHumanRole.cs), [here](https://github.com/BHoM/BHoM/blob/master/Humans_oM/Employee.cs), and [here](https://github.com/BHoM/BHoM/blob/master/Humans_oM/Skeleton.cs)), you will see I had already started a similar light ECS infrastructure with the concept of `Role` (only `IHumanRole` used at the moment). The idea of the name is that you attach extra functionality to a possibly varied set of objects so you give extra capability/purpose/role to that object. Not necessarily the most adequate name for a general purpose but something in between `BHoMExtendedProperties` and `Role` would be nice.
- There is a need for a discussion around the parts of ECS we want to embrace and the ones we want to leave behind. More specifically, the object definition aspect is only the tip of the iceberg while there is a good number of technical aspects to navigate around the handling (accessing, setting, updating,...) of those extended properties/roles.

IsakNaslundBh commented:
Not exactly the same, but linked to ECS, so might be something to keep in mind as well are the discussions started here:

https://github.com/BHoM/BHoM/issues/393

FraserGreenroyd commented:
@IsakNaslundBh yes good to link that in though I'd hope we could move this one forward quicker as it's holding up things from BEnv side whereas I'm not sure if materials is holding anything up for either side? Correct me if I'm wrong though cause I can't guess for your side :smile:

@adecler I have removed the `BHoMExtendedProperties` file and removed the `ElementID` from the interface (and moved it to the `BuildingElementExtendedProperties` for now). I haven't renamed anything yet as am waiting to see if anyone has any good ideas on how to name it :smile:

epignatelli commented:
What about `EnvironmentAware` ?

IsakNaslundBh commented:
@FraserGreenroyd no rush at all from the structures side atm, so no reason to hold this up for that sake! Just linking over as the principle overlap!

al-fisher commented:
@FraserGreenroyd @michaldengusiak As discussed suggesting you remove the `IBHoMExtendedProperties` and the Implementation on the `BH.oM.Base.BHoMObject` - such that this PR can therefore be a prototype of ECS simply for the `BuildingEnvironmentObjects`.

This will allow you to proceed and prototype in parallel to the `Human` example.   
And I would suggest @IsakNaslundBh we also discuss implementing a similar (or not) ECS approach for `Material`.

There are a number of effects and subtleties that will be uncovered - we will learn a lot.
In addition some further considerations will need to be considered around how to expose and reflect all ECS properties and behaviours in UI etc. Good to test in discipline code under specific use-cases before working to generalise and mapping to parent BHoM object and all dependant objects.

@adecler does this make sense to you too?

al-fisher commented:
Still struggling to think of the right term - somewhere between Role and ExtendedProperties although I do like what you are trying to achieve with BuildingAware @epignatelli 
Will keep thinking

FraserGreenroyd commented:
@al-fisher changes made as discussed :smile:

al-fisher commented:
Summary of discussions with @FraserGreenroyd @michaldengusiak @ZiolkowskiJakub -

@FraserGreenroyd _et al._ to post a first draft of the list (maybe just part of?! as previously super long) of analysis/design properties broken up and categorised into "Extended Properties" modules. Grouping the properties based on themes. This will help to flesh out the immediate BEnv implementation of an ECS prototype.

Also from discussion likely the name `BuildingElementExtendedProperties` will be changed / such a generic object removed - with then a number of `ExtendedProperties` objects created more specifically named around the classification of Properties it is containing. i.e. `CoolingLoadExtendedProperties` or `TemperatureExtendedProperties` etc.

Cheers all - great chat 😄 Please check the above makes sense/accurate - can follow up tomorrow too perhaps. 
@rwemay FYI

FraserGreenroyd commented:
As an FYI for all involved - this PR won't rebase nicely following changes made in PR #405 due to conflicts in the `Environment_oM.csproj` file. I'll be resolving this once @michaldengusiak finishes the diagram for us all but just to give everyone a heads up!
